core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala

```diff
@@ -81,6 +82,43 @@ private[spark] trait ExecutorAllocationClient {
     countFailures: Boolean,
     force: Boolean = false): Seq[String]
 
+  /**
+   * Request that the cluster manager decommission the specified executors.
+   * Default implementation delegates to kill, scheduler must override
+   * if it supports graceful decommissioning.
+   *
+   * @param executors identifiers of executors & decom info.
+   * @param adjustTargetNumExecutors whether the target number of executors will be adjusted down
+   *                                 after these executors have been decommissioned.
+   * @return the ids of the executors acknowledged by the cluster manager to be removed.
+   */
+  def decommissionExecutors(
```

Comment:
So I think we'd rather have a convenience method for the single implementation. Decommissioning one executor at a time could lead to a cascading migration, so if we've got a chunk of executors to decommission we'd rather encourage cluster managers to process that message as a block which is why I've implemented it this way.

Reply:
Makes sense. Please consider marking this single executor decom to be final so that it is clear that it is a helper method using an overrideable API. I agree with you that the default interface should be the bulk one.